crates/services/relayer/src/ports.rs

```diff
@@ -1,27 +1,133 @@
-    fn set_finalized_da_height(&mut self, block: DaBlockHeight) -> StorageResult<()>;
+    /// Set finalized da height that represent last block from da layer that got finalized.
+    /// This will only set the value if it is greater than the current.
+    fn set_finalized_da_height_to_at_least(
+        &mut self,
+        block: &DaBlockHeight,
+    ) -> StorageResult<()>;
 
-    /// Assume it is always set as initialization of database.
+    /// Get finalized da height that represent last block from da layer that got finalized.
+    /// Panics if height is not set as of initialization of database.
     fn get_finalized_da_height(&self) -> StorageResult<DaBlockHeight>;
 }
+
+impl<T, Storage> RelayerDb for T
+where
+    T: Send + Sync,
+    T: Transactional<Storage = Storage>,
+    T: StorageMutate<RelayerMetadata, Error = StorageError>,
+    Storage: StorageMutate<Messages, Error = StorageError>
+        + StorageMutate<RelayerMetadata, Error = StorageError>,
+{
+    fn insert_messages(&mut self, messages: &[CheckedMessage]) -> StorageResult<()> {
+        // A transaction is required to ensure that the height is
+        // set atomically with the insertion based on the current
+        // height. Also so that the messages are inserted atomically
+        // with the height.
+        let mut db_tx = self.transaction();
+        let db = db_tx.as_mut();
+
+        let mut max_height = None;
+        for message in messages {
+            db.storage::<Messages>()
+                .insert(message.id(), message.message())?;
+            let max = max_height.get_or_insert(0u64);
+            *max = (*max).max(message.message().da_height.0);
+        }
+        if let Some(height) = max_height {
+            grow_monotonically(db, &height.into())?;
+        }
+        db_tx.commit()?;
+        Ok(())
+    }
+
+    fn set_finalized_da_height_to_at_least(
+        &mut self,
+        height: &DaBlockHeight,
+    ) -> StorageResult<()> {
+        // A transaction is required to ensure that the height is
+        // set atomically with the insertion based on the current
+        // height.
+        let mut db_tx = self.transaction();
+        let db = db_tx.as_mut();
+        grow_monotonically(db, height)?;
+        db_tx.commit()?;
+        Ok(())
+    }
+
+    fn get_finalized_da_height(&self) -> StorageResult<DaBlockHeight> {
+        Ok(*StorageAsRef::storage::<RelayerMetadata>(&self)
+            .get(DA_HEIGHT_KEY)?
+            .ok_or(not_found!("DaBlockHeight missing for relayer"))?)
+    }
+}
+
+fn grow_monotonically<Storage>(
+    s: &mut Storage,
+    height: &DaBlockHeight,
+) -> StorageResult<()>
+where
+    Storage: StorageMutate<RelayerMetadata, Error = StorageError>,
+{
+    let current = (&s)
+        .storage::<RelayerMetadata>()
+        .get(DA_HEIGHT_KEY)?
+        .map(|cow| cow.as_u64());
+    match current {
+        Some(current) => {
+            if **height > current {
```

Comment:
Hmm, maybe we want to return an error here, if it is not monotonical

Reply:
Why would we error? If we are resyncing this could be true right?

Reply:
Is not the purpose of this function to return an error if the height is not monotonic? I guess with the error, we can be sure that each commit is monatomic.

If we do re-sync of the relayer, I guess we will clean up the state of the relayer first and start re-sync after.